refactor(sdk-coin-atom): only decode sequence from raw atom tx

michael-gallegos-bg opened the pull request:
We found that we no longer need the accountNumber or the chainId to properly construct a raw tx.

Also adds more testing regarding transactionBuilder and factory building.

TICKET: BG-67573

<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

jicli commented on modules/sdk-coin-atom/src/lib/transaction.ts:
```diff
@@ -125,6 +125,8 @@ export class Transaction extends BaseTransaction {
   }
 
   serialize(): string {
+    // TODO - Check if we can get the pub key from the encoded/decoded tx data
```
add a ticket for it plz

michael-gallegos-bg replied:
this is a draft PR, consulting with cosmos team on how we wish to proceed

michael-gallegos-bg commented:
From @jicli : "but we just changed to sequence id?"
Yes, so we used to implement this using the Stargate Client from the Cosmos JS SDK, which requires an account number/chain ID as inputs. However, we do not need this as we are using TSS, so only need the sequence number.
I refactored the code to account for this change.

mukeshsp commented on modules/sdk-coin-atom/src/lib/transactionBuilder.ts:
```diff
@@ -153,14 +155,14 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
+    //   if (!amount.denom) {
+    //     throw new BuildTransactionError('Invalid denom: undefined');
+    //   }
+    // }
```
Please remove the commented code and add the ticket number in todo.

michael-gallegos-bg replied:
Marking as resolved

mukeshsp commented on modules/sdk-coin-atom/src/lib/utils.ts:
```diff
@@ -203,17 +197,17 @@ export class Utils implements BaseUtils {
-    const stdSignature = encodeSecp256k1Signature(fromHex(pubkey), signatureBytes);
+  createSignedTxRaw(atomSignature: Signature, signDoc: SignDoc): TxRaw {
+    console.log('atom sig at this point: ' + atomSignature.signature);
+    console.log('hex of that signature: ' + atomSignature.signature.toString('hex'));
```
remove console log

michael-gallegos-bg replied:
Marking as resolved

mukeshsp commented on modules/sdk-coin-atom/src/lib/utils.ts:
```diff
@@ -203,17 +197,17 @@ export class Utils implements BaseUtils {
+    console.log('atom sig at this point: ' + atomSignature.signature);
+    console.log('hex of that signature: ' + atomSignature.signature.toString('hex'));
+    const stdSignature = encodeSecp256k1Signature(fromHex(atomSignature.publicKey.pub), atomSignature.signature);
+    console.log('stdSig: ' + stdSignature.signature);
```
remove console

michael-gallegos-bg replied:
Marking as resolved

Vijay-Jagannathan commented on modules/sdk-coin-atom/src/lib/utils.ts:
```diff
@@ -225,6 +219,31 @@ export class Utils implements BaseUtils {
   }
+
+  isSignedRawTx(rawTransaction: string): boolean {
+    // if (!this.isValidRawTransaction(rawTransaction)) {
+    //   return false;
+    // }
```
Good to remove

Vijay-Jagannathan commented on modules/sdk-coin-atom/src/lib/utils.ts:
```diff
@@ -225,6 +219,31 @@ export class Utils implements BaseUtils {
+    return false;
+  }
+
+  getSignerInfoFromRawSignedTx(rawTransaction: string) {
+    if (!this.isSignedRawTx(rawTransaction)) {
```
This method gets called within `fromRawTransaction` in the transaction class. Before this method is called, we already check if the transaction is indeed signed or not (Line 123 in transaction class), so we might not need line 235 (unless this method can be called from elsewhere where we need to check if the transaction is signed)

mukeshsp replied:
Let's keep this function for now, as we may call it in another class in the future!

michael-gallegos-bg replied:
Good call - I was trying to explicitly show in this version that currently we are only supporting signed tx data, and are working on unsigned with the HSM team

sachushaji commented on modules/sdk-coin-atom/package.json:
```diff
@@ -44,7 +44,6 @@
     "@bitgo/statics": "^14.1.0",
     "@bitgo/utxo-lib": "^7.5.0",
     "@cosmjs/amino": "^0.29.5",
-    "@cosmjs/crypto": "^0.29.5",
```
No package lock changes?

michael-gallegos-bg replied:
Isn't the lock-file automatically generated?
I do not see it in other coin modules
https://github.com/BitGo/BitGoJS/tree/master/modules/sdk-coin-atom
https://github.com/BitGo/BitGoJS/tree/master/modules/sdk-coin-sol

sachushaji replied:
Yeah, Lock files are autogenerated when you do `yarn`  again if you made changes to dependencies. I checkout locally and seems like it's a peer dependencies for some of the packages like `@cosmjs/amino` etc, probably explains why lock file was not modified.

michael-gallegos-bg replied:
Oh yes - they're dependencys of each other, thanks for checking!

sachushaji commented on modules/sdk-coin-atom/src/lib/iface.ts:
```diff
@@ -29,7 +29,7 @@ export interface TxData extends AtomTransaction {
   type: string;
   signerAddress: string;
-  explicitSignerData: SignerData;
+  sequence: number;
```
Seems to be breaking change?, mention in the commit message ([format](https://www.conventionalcommits.org/en/v1.0.0/#:~:text=in%20Semantic%20Versioning\).-,BREAKING,-CHANGE%3A%20a))

michael-gallegos-bg replied:
Updated the commit message for the breaking change commit!
